File: benchmarks/medium/M3-hashmap-stress/fsharp/Program.fs
open System
open System.Collections.Generic
open System.Diagnostics

[<EntryPoint>]
let main _ =
    let n = 5_000_000
    let lookups = 1_000_000
    let runs = 10
    let insertTimings = Array.zeroCreate<float> runs
    let lookupTimings = Array.zeroCreate<float> runs

    printfn "=== BENCHMARK: M3 - Hash Map Stress Test ==="
    printfn "Language: F#"
    printfn "Runtime:  .NET %O" Environment.Version
    printfn ""

    let mutable verifySum = 0L
    let sw = Stopwatch()

    for run in 0 .. runs - 1 do
        // Insert phase
        sw.Restart()
        let map = Dictionary<string, int64>(n)
        for i in 0 .. n - 1 do
            map.[sprintf "user_%d" i] <- int64 i * 31L + 7L
        sw.Stop()
        insertTimings.[run] <- sw.Elapsed.TotalMilliseconds

        // Lookup phase
        sw.Restart()
        let mutable state = 42u
        let mutable sum = 0L
        for _ in 0 .. lookups - 1 do
            state <- state * 1664525u + 1013904223u
            let idx = int (uint64 state % uint64 n)
            sum <- sum + map.[sprintf "user_%d" idx]
        sw.Stop()
        lookupTimings.[run] <- sw.Elapsed.TotalMilliseconds

        verifySum <- sum
        printfn "Run %2d: Insert: %8.2f ms | Lookup: %8.2f ms | Total: %8.2f ms"
            (run + 1) insertTimings.[run] lookupTimings.[run] (insertTimings.[run] + lookupTimings.[run])

    printfn ""

    let printStats label (timings: float[]) =
        let minT = Array.min timings
        let maxT = Array.max timings
        let avg = Array.average timings
        let variance = timings |> Array.averageBy (fun t -> (t - avg) ** 2.0)
        let stddev = sqrt variance
        printfn "%s:" label
        printfn "  Min:    %.2f ms" minT
        printfn "  Avg:    %.2f ms" avg
        printfn "  Max:    %.2f ms" maxT
        printfn "  StdDev: %.2f ms" stddev
        printfn ""

    printStats "Insert" insertTimings
    printStats "Lookup" lookupTimings

    let totalTimings = Array.init runs (fun i -> insertTimings.[i] + lookupTimings.[i])
    printStats "Total" totalTimings

    printfn "Verification: sum = %d" verifySum
    0

File: benchmarks/medium/M3-hashmap-stress/csharp/Program.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

const int N = 5_000_000;
const int LOOKUPS = 1_000_000;
const int RUNS = 10;

double[] insertTimings = new double[RUNS];
double[] lookupTimings = new double[RUNS];

Console.WriteLine("=== BENCHMARK: M3 - Hash Map Stress Test ===");
Console.WriteLine("Language: C#");
Console.WriteLine($"Runtime:  .NET {Environment.Version}");
Console.WriteLine();

long verifySum = 0;
Stopwatch sw = new();

for (int run = 0; run < RUNS; run++)
{
    // Insert phase
    sw.Restart();
    var map = new Dictionary<string, long>(N);
    for (int i = 0; i < N; i++)
    {
        map[$"user_{i}"] = (long)i * 31 + 7;
    }
    sw.Stop();
    insertTimings[run] = sw.Elapsed.TotalMilliseconds;

    // Lookup phase
    sw.Restart();
    uint state = 42u;
    long sum = 0;
    for (int j = 0; j < LOOKUPS; j++)
    {
        state = unchecked(state * 1664525u + 1013904223u);
        int idx = (int)(state % (uint)N);
        sum += map[$"user_{idx}"];
    }
    sw.Stop();
    lookupTimings[run] = sw.Elapsed.TotalMilliseconds;

    verifySum = sum;
    Console.WriteLine($"Run {run + 1,2}: Insert: {insertTimings[run],8:F2} ms | Lookup: {lookupTimings[run],8:F2} ms | Total: {insertTimings[run] + lookupTimings[run],8:F2} ms");
}

Console.WriteLine();
PrintStats("Insert", insertTimings);
PrintStats("Lookup", lookupTimings);

double[] totalTimings = new double[RUNS];
for (int i = 0; i < RUNS; i++)
    totalTimings[i] = insertTimings[i] + lookupTimings[i];
PrintStats("Total", totalTimings);

Console.WriteLine($"Verification: sum = {verifySum}");

static void PrintStats(string label, double[] timings)
{
    double min = timings[0], max = timings[0], sum = 0;
    foreach (double t in timings)
    {
        sum += t;
        if (t < min) min = t;
        if (t > max) max = t;
    }
    double avg = sum / timings.Length;
    double variance = 0;
    foreach (double t in timings)
        variance += (t - avg) * (t - avg);
    variance /= timings.Length;
    double stddev = Math.Sqrt(variance);

    Console.WriteLine($"{label}:");
    Console.WriteLine($"  Min:    {min:F2} ms");
    Console.WriteLine($"  Avg:    {avg:F2} ms");
    Console.WriteLine($"  Max:    {max:F2} ms");
    Console.WriteLine($"  StdDev: {stddev:F2} ms");
    Console.WriteLine();
}
